### PR TITLE
Making endpoints and addresses easier

### PR DESCRIPTION
**Pull request by rynowak**
This makes endpoints and addresses easier to work with by dropping the
'metadata first' approach for the the things that are really at the core
of the dispatcher.

### CODE REVIEW

**Comment by rynowak** [comment]
@davidfowl thoughts?

/cc @jbagga please have a look as well

File: src/Microsoft.AspNetCore.Dispatcher/ITemplateAddress.cs [hunk 1]
```diff
@@ -3,8 +3,10 @@
 
 namespace Microsoft.AspNetCore.Dispatcher
 {
-    public interface IDispatcherValueSelectableEndpoint
+    public interface ITemplateAddress
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/ITemplateAddress.cs]: What do you need to generate a URL?

File: src/Microsoft.AspNetCore.Dispatcher/ITemplateEndpoint.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Dispatcher
+{
+    public interface ITemplateEndpoint
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/ITemplateEndpoint.cs]: What do you need to build a route table?

File: src/Microsoft.AspNetCore.Dispatcher/ITemplateEndpoint.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    {
+        string HttpMethod { get; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/ITemplateEndpoint.cs]: We could also put this on another interface, but it feels like a core concept to me

Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/ITemplateEndpoint.cs]: It's not strictly required

File: src/Microsoft.AspNetCore.Dispatcher/TemplateEndpoint.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Dispatcher
+{
+    public class TemplateEndpoint : Endpoint, ITemplateEndpoint
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/TemplateEndpoint.cs]: There are a lot of constructors here. Don't think of this as something that's going to be in the face of most users. We will have a more polished startup experience.

File: src/Microsoft.AspNetCore.Routing/Dispatcher/RouteTemplateUrlGenerator.cs [hunk 5]
```diff
@@ -37,14 +37,13 @@ public string GenerateUrl(HttpContext httpContext, object values)
                 throw new ArgumentNullException(nameof(values));
             }
 
-            var address = _addressSelector.SelectAddress(new DispatcherValueCollection(values));
+            var address = _addressSelector.SelectAddress(new DispatcherValueCollection(values)) as ITemplateAddress;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Dispatcher/RouteTemplateUrlGenerator.cs]: Contract of the `TemplateAddressSelector` returns an `ITemplateAddress + Address`. We don't have sum types tho. Sad panda

File: src/Microsoft.AspNetCore.Routing/Dispatcher/TreeDispatcher.cs [hunk 6]
```diff
@@ -142,16 +142,16 @@ private Cache CreateCache()
...
                 {
                     continue;
                 }
 
-                if (!groups.TryGetValue(new Key(metadata.Order, metadata.RouteTemplate), out var group))
+                if (!groups.TryGetValue(new Key(0, templateEndpoint.Template), out var group))
                 {
                     group = new List<Endpoint>();
-                    groups.Add(new Key(metadata.Order, metadata.RouteTemplate), group);
+                    groups.Add(new Key(0, templateEndpoint.Template), group);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Dispatcher/TreeDispatcher.cs]: ignoring order for now, I think we can get rid of that concept

Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing/Dispatcher/TreeDispatcher.cs]: By making it separate dispatchers with a single tree

File: src/Microsoft.AspNetCore.Dispatcher/TemplateEndpoint.cs (hunk 4 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Dispatcher/TemplateEndpoint.cs]: Maybe a builder pattern would be better.

File: src/Microsoft.AspNetCore.Dispatcher/ITemplateEndpoint.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+        string Template { get; }
+
+        DispatcherValueCollection Values { get; }
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Dispatcher/ITemplateEndpoint.cs]: Still not sure why this is available on the end point

File: samples/DispatcherSample/Startup.cs [hunk 8]
```diff
@@ -32,27 +31,27 @@ public void ConfigureServices(IServiceCollection services)
...
-                        new SimpleEndpoint(Home_About, new object[] { new RouteTemplateMetadata("{controller=Home}/{action=Index}/{id?}"), }, new { controller = "Home", action = "About", }, "Home:About()"),
-                        new SimpleEndpoint(Admin_Index, new object[] { new RouteTemplateMetadata("{controller=Home}/{action=Index}/{id?}"), }, new { controller = "Admin", action = "Index", }, "Admin:Index()"),
-                        new SimpleEndpoint(Admin_GetUsers, new object[] { new RouteTemplateMetadata("{controller=Home}/{action=Index}/{id?}"), new HttpMethodMetadata("GET"), new AuthorizationPolicyMetadata("Admin"), }, new { controller = "Admin", action = "Users", }, "Admin:GetUsers()"),
-                        new SimpleEndpoint(Admin_EditUsers, new object[] { new RouteTemplateMetadata("{controller=Home}/{action=Index}/{id?}"),  new HttpMethodMetadata("POST"), new AuthorizationPolicyMetadata("Admin"), }, new { controller = "Admin", action = "Users", }, "Admin:EditUsers()"),
+                        new TemplateEndpoint("{controller=Home}/{action=Index}/{id?}", new { controller = "Home", action = "Index", }, Home_Index, "Home:Index()"),
```
Comment (davidfowl) [inline samples/DispatcherSample/Startup.cs]: Why are we still specifying defaults?

File: src/Microsoft.AspNetCore.Dispatcher/TemplateEndpoint.cs (hunk 4 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/TemplateEndpoint.cs]: I built two different versions of that yesterday. I wanted to understand your ideas about application model before going further with it.

File: src/Microsoft.AspNetCore.Dispatcher/ITemplateEndpoint.cs (hunk 7 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/ITemplateEndpoint.cs]: These are the values that need to be present to select the endpoint. We can go further here by just doing token replacement

Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/ITemplateEndpoint.cs]: Even if we do token replacement this property still has to be here. Attribute routing treats these like defaults, that's required for ambient values to work

File: samples/DispatcherSample/Startup.cs (hunk 8 above)
Comment (rynowak) [reply samples/DispatcherSample/Startup.cs]: Even if we do token replacement this property still has to be here. Attribute routing treats these like defaults, that's required for ambient values to work

**Comment by rynowak** [comment]
Discussed this in person, I'm going to go ahead and merge for now since this is cleanup.

**Pull request merged**